network/requestTracker.go

```diff
@@ -0,0 +1,385 @@
+// Copyright (C) 2019 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package network
+
+import (
+	"fmt"
+	"net"
+	"net/http"
+	"sort"
+	"time"
+
+	"github.com/algorand/go-deadlock"
+
+	"github.com/algorand/go-algorand/config"
+	"github.com/algorand/go-algorand/logging"
+	"github.com/algorand/go-algorand/logging/telemetryspec"
+)
+
+const (
+	// maxHeaderReadTimeout is the time limit where items would remain in the acceptedConnections cache before being pruned.
+	// certain malicious connections would never get to the http handler, and therefore must be pruned every so often.
+	maxHeaderReadTimeout = 30 * time.Second
+)
+
+// TrackerRequest hold the tracking data associated with a single request.
+type TrackerRequest struct {
+	created            time.Time
+	remoteHost         string
+	remotePort         string
+	remoteAddr         string
+	request            *http.Request
+	otherTelemetryGUID string
+	otherInstanceName  string
+	otherPublicAddr    string
+}
+
+// makeTrackerRequest creates a new TrackerRequest.
+func makeTrackerRequest(remoteAddr, remoteHost, remotePort string, createTime time.Time) *TrackerRequest {
+	if remoteHost == "" {
+		remoteHost, remotePort, _ = net.SplitHostPort(remoteAddr)
+	}
+
+	return &TrackerRequest{
+		created:    createTime,
+		remoteAddr: remoteAddr,
+		remoteHost: remoteHost,
+		remotePort: remotePort,
+	}
+}
+
+// hostIncomingRequests holds all the requests that are originating from a single host.
+type hostIncomingRequests struct {
+	remoteHost string
+	requests   []*TrackerRequest // this is an ordered list, according to the requestsHistory.created
```

Comment:
could this create a memory problem if there are too many requests from a single host?

Reply:
It's not being stored for a long period of time. In fact, it's being stored only for the duration of the window size - which defaults to 1 second.